theano/sandbox/cuda/dnn.py

```diff
@@ -1340,41 +1343,95 @@ def c_compiler(self):
     def do_constant_folding(self, node):
         return False
 
-    def __init__(self, ws=(1, 1), stride=(1, 1), mode='max', pad=(0, 0)):
+    def __init__(self, ws=(1, 1), stride=(1, 1), mode='max',
+                 pad=(0, 0), ndim=None):
+        nd = None
+        if isinstance(ws, (tuple, list)):
+            nd = len(ws)
+        elif isinstance(stride, (tuple, list)):
+            nd = len(stride)
+        elif isinstance(pad, (tuple, list)):
+            nd = len(pad)
+        if ndim is None:
+            if nd is None:
+                raise RuntimeError(
+                    "Not able to detect if we do 2d or 3d pooling")
+            else:
+                ndim = nd
+        elif nd is not None:
+            assert ndim == nd
         if mode == 'average':
             mode = 'average_inc_pad'
         assert mode in ('max', 'average_inc_pad', 'average_exc_pad')
         self.mode = mode
+        assert ndim in [2, 3]
+        self.ndim = ndim
 
-        assert len(ws) == len(stride) and len(stride) == len(pad)
-        assert len(ws) in (2, 3)
-        self.ws = ws
-        self.stride = stride
-        self.pad = pad
-
-        if (pad[0] != 0 or pad[1] != 0) and version() == -1:
-            raise RuntimeError("CuDNN pooling with padding requires CuDNN v2")
         if self.get_ndim() == 3 and version() < (3000, 3000):
             raise RuntimeError("CuDNN 3d pooling requires CuDNN v3")
 
     def get_ndim(self):
-        return len(self.ws)
+        return self.ndim
+
+    def get_ws(self, node):
+        if hasattr(self, "ws"):
+            return self.ws
+        else:
+            return [node.inputs[0][i] for i in range(self.ndim)]
+
+    def get_stride(self, node):
+        if hasattr(self, "stride"):
+            return self.stride
+        else:
+            return [node.inputs[1][i] for i in range(self.ndim)]
+
+    def get_pad(self, node):
+        if hasattr(self, "pad"):
+            return self.pad
+        else:
+            return [node.inputs[2][i] for i in range(self.ndim)]
 
     def __setstate__(self, d):
         self.__dict__.update(d)
-        if not hasattr(self, 'pad'):
+        if not hasattr(self, 'pad') and hasattr(self, 'ws'):
             self.pad = (0, 0)
 
-    def make_node(self):
-        if self.pad != (0, 0) and version() == -1:
-            raise RuntimeError("CuDNN pooling with padding requires CuDNN v2")
-
-        return Apply(self, [],
+    def make_node(self, ws, stride, pad):
+        ws = tensor.as_tensor_variable(ws)
+        stride = tensor.as_tensor_variable(stride)
+        pad = tensor.as_tensor_variable(pad)
+        assert ws.ndim == stride.ndim == pad.ndim == 1
+        ws = ws.astype(theano.scalar.upcast("int32", ws.dtype))
+        stride = stride.astype(theano.scalar.upcast("int32", stride.dtype))
+        pad = pad.astype(theano.scalar.upcast("int32", pad.dtype))
+        # TODO how to detect that int are int64 or int32?
```

Comment:
You could use the 'intp' dtype.  That is the int type which is the same size as ssize_t for numpy.